<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good) Exercises</span></div>

# Exercise 1

In [ ]:
%%timeit #Do not print result , timeit runs the cell multiple times 
#(-n 10): 10 loops, (-r 1): 1 run, all together 1 run of 10 loops each. syntax: %%timeit -n 10 -r 1
odd_list = []
for num in range(0,10):
    if num % 2 == 0:
        continue
    else:
        odd_list.append(num**2)
#print(odd_list)

1.61 μs ± 0 ns per loop (mean ± std. dev. of 1 run, 10 loops each)


# Exercise 2

In [ ]:
%%timeit 
new_odd_list = [num**2 for num in range(0, 10) if num % 2 == 1]

1.03 μs ± 0 ns per loop (mean ± std. dev. of 1 run, 10 loops each)


A list comprehension is faster than a for loop. A list comprehension took an average of 1.03 μs ± 0 ns per loop while a for loop took an average of 1.61 μs ± 0 ns per loop.

# Exercise 4

In [23]:
quantity = 100
half_original_amount = quantity / 2
num_years = 0
while quantity > half_original_amount:
    quantity *= 0.95
    num_years += 1

print(f"It takes {num_years} years and the quantity left is {round(quantity, 2)} units rounded off to 2 decimal places.")

It takes 14 years and the quantity left is 48.77 units rounded off to 2 decimal places.


# Exercise 5

In [ ]:
co2_data = [
    (2000, 369.55), (2001, 371.14), (2002, 373.28), 
    (2003, 375.80), (2004, 377.52), (2005, 379.80), 
    (2006, 381.90), (2007, 383.79), (2008, 385.60), 
    (2009, 387.43), (2010, 389.90), (2011, 391.65), 
    (2012, 393.85), (2013, 396.52), (2014, 398.65),
    (2015, 400.83), (2016, 404.24), (2017, 406.55), 
    (2018, 408.52), (2019, 411.44), (2020, 414.24)
]
data_length = len(co2_data)
for index in range(1, data_length):
    year, CO2_level = co2_data[index]
    previous_year_CO2_level = co2_data[index - 1][1]
    diff_in_ppm = CO2_level - previous_year_CO2_level
    if diff_in_ppm >= 3:
        print(f"Year: {year}, change in concentration: {round(diff_in_ppm, 2)}ppm rounded off to 2 decimal places")

Year: 2016, change in concentration: 3.41ppm rounded off to 2 decimal places
